### ISSUE TITLE
Error: fullwidth low line (U+FF3F) is not valid in identifier

### ISSUE THREAD

**Issue by jplaisted** [google/closure-compiler#2851]
The fullwidth low line (U+FF3F) character should be valid to use in a Javascript identifier but the compiler throws a parse error. [Example](https://closure-compiler-debugger.appspot.com/#input0%3Dclass%2520MyClass%2520%257B%2520constructor()%2520%257B%2520this.foo%25EF%25BC%25BF1%253B%2520%257D%2520%257D%250A%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3Dtrue%26CHECK_SYMBOLS%3Dtrue%26MISSING_PROPERTIES%3Dtrue%26TRANSPILE%3Dtrue%26CHECK_TYPES%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue).

The spec states that:
> any Unicode code point with the Unicode property “ID_Continue”, “Other_ID_Continue”, or “Other_ID_Start”
should be valid.

**Comment by brad4d** [comment]
The trouble with making the parser match the ES spec for valid identifier characters is the parser has to work when converted from Java to JavaScript by GWT.

The GWT version of the Java `Character` class is missing the methods that we need, and the methods it actually does have only work correctly for ASCII.

gwtproject/gwt#1989

Work on that issue seems to have stalled in 2015.

We have a workaround in place for one special character that gets used by Angular.
https://github.com/google/closure-compiler/blob/9dfc4e72976d4ca6141a0b3f1ece959aa027ef8a/src/com/google/javascript/jscomp/parsing/parser/Scanner.java#L796

@jplaisted what was the use-case that required fullwidth low line (U+FF3F)?
Maybe we should add another special case for that.

**Comment by brad4d** [comment]
@gkdn this bug was filed because the fullwidth low line character is used in your project.
I figured you'd want to know about it.

**Comment by gkdn** [comment]
You mean issue is Character.isLetter on handles ASCII?

**Comment by brad4d** [comment]
Actually, I think the central issue is that `Character` in GWT doesn't implement the methods we really should be using `isUnicodeIdentifierStart()` and `isUnicodeIdentifierPart()`. It seems the JS spec defines identifiers as those having particular unicode properties that I think correspond to those methods.

https://tc39.github.io/ecma262/#prod-IdentifierPart

But, yeah, it's also a problem that `Character.isLetter()` is only correct for ASCII.

**Comment by brad4d** [comment]
This issue came up because the J2CL minifier generates code containing the "fullwidth low line" character.
It is not expected that such code should be fed to closure-compiler, but @jplaisted ran into a corner case where it happened.

He worked around it.

**Comment by jplaisted** [comment]
Yeah, should've explained, sorry. I just saw the compiler didn't technically meet the specification and filed a bug. Not a huge priority, but would be nice if it was fixed.

**Comment by gkdn** [comment]
If this is happening on j2cl minified source for processing for ES3/ES5 transpilation on the processor, that seems like a valid use case. Maybe we can update the minifier for that as well.
